File: radixsort.js
/**
 * --> radixsort -->
 * loop on all items, group by last bit
 * loop on all items, group by 2nd from the end bit
 * etc.
 * [1000,10,1,11]
 */
function radixsort(array, bitsOfKeys) {
  if (!array || array.length < 2) {
    return array;
  }
  for (var bitIndex = 0; bitIndex < bitsOfKeys; bitIndex++) {
    var mask = 1 << bitIndex;
    var firstUpperIndex = 0;
    for (var i = 0; i < array.length; i++) {
      if (!(mask & array[i])) {
        swap(array, i, firstUpperIndex);
        firstUpperIndex++;
      }
    }
  }
  return array;
}

function swap(array, i, j) {
  var temp = array[i];
  array[i] = array[j];
  array[j] = temp;
}

console.log(radixsort([8,2,1,3], 8));

File: RadixSort.java
import java.util.*;

public class RadixSort {
  public static void main(String[] args) {
    int[] array = new int[]{78,22,11,1,2,3,4,9,8,7,6,5,53,77};
    int[] sorted = radixSort(array, 8);
    System.out.println(Arrays.toString(sorted));
  }

  /** Radix sort, in place. */
  public static int[] radixSort(int[] array, int bits) {
    if (array == null || array.length < 2) {
      return array;
    }

    for (int pos = 0; pos < bits; pos++) {
      int mask = 1 << pos;
      int zerosOnLeft = 0;
      for (int i = 0; i < array.length; i++) {
        if ((array[i] & mask) == 0) {
          // Move it before 1s.
          move(array, i, zerosOnLeft);
          zerosOnLeft++;
        }
      }
    }
    return array;
  }

  /** Move item on j to index i shifting items in between to the right. j >= i */  
  private static void move(int[] array, int j, int i) {
    if (j <= i) {
      return;
    }
    for (int toMove = j - 1; toMove >= i; toMove--) {
      swap(array, toMove, toMove + 1);
    }
  }
  
  private static void swap(int[] array, int i, int j) {
    int temp = array[i];
    array[i] = array[j];
    array[j] = temp;
  }
}
